cloudsync/smartsync.py

```diff
@@ -257,6 +258,22 @@ def _ent_to_smartinfo(self, ent: Optional[SyncEntry], local_dir_info: Optional[D
                            )
         return retval
 
+    def _include_ent(self, ent: SyncEntry) -> bool:
+        '''Returns True if Smartsync should include the ent during listdir/info'''
+        if not ent:
+            return False
+
+        # Ent has been trashed locally or remotely and there is no local counterpart
+        if ent[LOCAL].exists in (TRASHED, MISSING) \
```

Comment:
Local existence check should hit the provider exists method, either instead of the check against the state table, or in addition as a double check

Reply:
The state table checks are to avoid expensive checks against the remote provider, local provider checks are assumed to be cheap, and should be used instead of the state table which can be out of date

Reply:
I know we had talked about local provider hits being cheap in comparison to remote provider hits and while I agree with that in principal, I think we may be taking it too far. We're already doing a local listdir to check the local state and I'm worried changing this to hit the provider as well could crush performance. For a large directory, this would be an `os.stat()` call for every file existing remotely.

Reply:
I think the upside is pretty minimal insofar as the fs can rectify this situation and should already prefer files that exist locally as is and the downside seems significant IMO

Reply:
The state table can be out of date, so it is important to check against the local provider. At least only check against the provider if the condition, as written, evaluates to True, as a double check before returning False there. Or pass in the directory entry as an optional argument, if you have it, or a simple Boolean argument if you know at the time that the method is called whether it exists or not... Just we *cannot* solely rely on the state table for information about the local file system.

Reply:
Preferring files that exist locally is a condition that you are assuming appears as part of the the condition in which this method is called... Is that true everywhere? If so, move that logic into this method, and if not, then you can't depend on that preference in every case

Reply:
I think the whole cloudsync problem becomes incredibly arduous if we accept the notion that we can't trust anything that is in the statedb. We trust what's in the statedb in many places throughout the code and yes we call `get_latest` occasionally but the information can always change post `get_latest` call so I see that as more a strategic way to reduce race conditions in certain situations as opposed to making cloudsync an all-knowing oracle. 

Regarding listdir, we already do a local listdir and prefer any local results that we get back. This is really just intended to be a function to filter out remote SyncEnts that shouldn't be returned back to the caller.

It is largely true that we are checking if the local file exists before hitting smartsync. For instance, in `fs_stat()` we don't hit smartsync unless `os.stat()` raises a `FileNotFoundError`. I think asking the filesystem to not introspect the files on disk is an unrealistic standard and I don't think demonstrates a good separation of responsibilities. I would prefer the filesystem to retain ownership of files on disk and the decisions around them. I think this provides a better separation of responsibilities and more flexibility going forward.

Reply:
The statedb is delayed and that can't be gotten around by wishful thinking. When saving, MSoffice creates a temp file, and then deletes the original file and renames the temp over the original. The original is now `TRASHED` in the statedb, but we can't do a `smart_info_path()` on that file because the new file hasn't shown up in the statedb yet, and we're not looking on disk to see if it's there. If you can't immediately get the file info for a file you just created, then that is a bug

Reply:
It the file exists locally it should work fine. We're only filtering out the remote entry